File: ios/Sources/Modules/NFCModule.swift
/**
 * NFC 模块 - 近场通信功能
 *
 * 提供 NFC 读写功能（CoreNFC）
 */

import Foundation
import CoreNFC

// MARK: - NFC 模块

@available(iOS 13.0, *)
public class NFCModule: NSObject, BridgeModule {
    
    public let moduleName = "NFC"
    public let methods = [
        "IsAvailable",
        "IsEnabled",
        "StartScan",
        "StopScan",
        "WriteTag"
    ]
    
    private weak var bridge: WebViewBridge?
    private var ndefSession: NFCNDEFReaderSession?
    private var tagSession: NFCTagReaderSession?
    private var pendingCallback: ((Result<Any?, BridgeError>) -> Void)?
    private var writeMessage: NFCNDEFMessage?
    private var isWriteMode = false
    
    public init(bridge: WebViewBridge) {
        self.bridge = bridge
        super.init()
    }
    
    public func handleRequest(
        method: String,
        params: [String: AnyCodable],
        callback: @escaping (Result<Any?, BridgeError>) -> Void
    ) {
        switch method {
        case "IsAvailable":
            isAvailable(callback: callback)
        case "IsEnabled":
            isEnabled(callback: callback)
        case "StartScan":
            startScan(params: params, callback: callback)
        case "StopScan":
            stopScan(callback: callback)
        case "WriteTag":
            writeTag(params: params, callback: callback)
        default:
            callback(.failure(BridgeError.methodNotFound("\(moduleName).\(method)")))
        }
    }
    
    // MARK: - IsAvailable
    
    /// 检查 NFC 是否可用
    private func isAvailable(callback: @escaping (Result<Any?, BridgeError>) -> Void) {
        let available = NFCNDEFReaderSession.readingAvailable
        
        callback(.success([
            "isAvailable": available,
            "ndefSupported": available,
            "tagSupported": NFCTagReaderSession.readingAvailable
        ]))
    }
    
    // MARK: - IsEnabled
    
    /// 检查 NFC 是否开启（iOS 上等同于 isAvailable）
    private func isEnabled(callback: @escaping (Result<Any?, BridgeError>) -> Void) {
        callback(.success([
            "isEnabled": NFCNDEFReaderSession.readingAvailable
        ]))
    }
    
    // MARK: - StartScan
    
    /// 开始扫描 NFC 标签
    private func startScan(
        params: [String: AnyCodable],
        callback: @escaping (Result<Any?, BridgeError>) -> Void
    ) {
        guard NFCNDEFReaderSession.readingAvailable else {
            callback(.failure(BridgeError(code: .featureDisabled, message: "NFC 不可用")))
            return
        }
        
        // 停止现有会话
        ndefSession?.invalidate()
        
        let alertMessage = params["alertMessage"]?.stringValue ?? "将设备靠近 NFC 标签"
        let keepSessionAlive = params["keepSessionAlive"]?.boolValue ?? false
        
        isWriteMode = false
        pendingCallback = callback
        
        ndefSession = NFCNDEFReaderSession(
            delegate: self,
            queue: nil,
            invalidateAfterFirstRead: !keepSessionAlive
        )
        ndefSession?.alertMessage = alertMessage
        ndefSession?.begin()
    }
    
    // MARK: - StopScan
    
    /// 停止扫描
    private func stopScan(callback: @escaping (Result<Any?, BridgeError>) -> Void) {
        ndefSession?.invalidate()
        ndefSession = nil
        tagSession?.invalidate()
        tagSession = nil
        
        callback(.success(["stopped": true]))
    }
    
    // MARK: - WriteTag
    
    /// 写入 NFC 标签
    private func writeTag(
        params: [String: AnyCodable],
        callback: @escaping (Result<Any?, BridgeError>) -> Void
    ) {
        guard NFCNDEFReaderSession.readingAvailable else {
            callback(.failure(BridgeError(code: .featureDisabled, message: "NFC 不可用")))
            return
        }
        
        // 构建 NDEF 消息
        var records: [NFCNDEFPayload] = []
        
        // 支持写入文本
        if let text = params["text"]?.stringValue {
            if let payload = NFCNDEFPayload.wellKnownTypeTextPayload(
                string: text,
                locale: Locale.current
            ) {
                records.append(payload)
            }
        }
        
        // 支持写入 URI
        if let uri = params["uri"]?.stringValue,
           let url = URL(string: uri) {
            if let payload = NFCNDEFPayload.wellKnownTypeURIPayload(url: url) {
                records.append(payload)
            }
        }
        
        // 支持写入自定义记录
        if let recordsData = params["records"]?.arrayValue {
            for recordInfo in recordsData {
                if let record = recordInfo as? [String: Any] {
                    if let payload = createNDEFPayload(from: record) {
                        records.append(payload)
                    }
                }
            }
        }
        
        guard !records.isEmpty else {
            callback(.failure(BridgeError.invalidParams("text/uri/records")))
            return
        }
        
        writeMessage = NFCNDEFMessage(records: records)
        isWriteMode = true
        pendingCallback = callback
        
        let alertMessage = params["alertMessage"]?.stringValue ?? "将设备靠近要写入的 NFC 标签"
        
        ndefSession?.invalidate()
        ndefSession = NFCNDEFReaderSession(
            delegate: self,
            queue: nil,
            invalidateAfterFirstRead: false
        )
        ndefSession?.alertMessage = alertMessage
        ndefSession?.begin()
    }
    
    // MARK: - 辅助方法
    
    private func createNDEFPayload(from record: [String: Any]) -> NFCNDEFPayload? {
        guard let typeString = record["type"] as? String,
              let payloadString = record["payload"] as? String,
              let payloadData = payloadString.data(using: .utf8) else {
            return nil
        }
        
        let tnf: NFCTypeNameFormat
        switch record["tnf"] as? String {
        case "empty":
            tnf = .empty
        case "wellKnown":
            tnf = .nfcWellKnown
        case "media":
            tnf = .media
        case "absoluteUri":
            tnf = .absoluteURI
        case "external":
            tnf = .nfcExternal
        default:
            tnf = .unknown
        }
        
        let typeData = typeString.data(using: .utf8) ?? Data()
        let identifier = (record["id"] as? String)?.data(using: .utf8) ?? Data()
        
        return NFCNDEFPayload(
            format: tnf,
            type: typeData,
            identifier: identifier,
            payload: payloadData
        )
    }
    
    private func parseNDEFMessage(_ message: NFCNDEFMessage) -> [[String: Any]] {
        return message.records.map { record in
            var info: [String: Any] = [
                "tnf": tnfString(record.typeNameFormat),
                "type": String(data: record.type, encoding: .utf8) ?? "",
                "identifier": String(data: record.identifier, encoding: .utf8) ?? ""
            ]
            
            // 解析 payload
            if let text = record.wellKnownTypeTextPayload().0 {
                info["text"] = text
                info["locale"] = record.wellKnownTypeTextPayload().1?.identifier ?? ""
            }
            
            if let url = record.wellKnownTypeURIPayload() {
                info["uri"] = url.absoluteString
            }
            
            // 原始 payload
            info["payload"] = record.payload.base64EncodedString()
            info["payloadText"] = String(data: record.payload, encoding: .utf8)
            
            return info
        }
    }
    
    private func tnfString(_ tnf: NFCTypeNameFormat) -> String {
        switch tnf {
        case .empty:
            return "empty"
        case .nfcWellKnown:
            return "wellKnown"
        case .media:
            return "media"
        case .absoluteURI:
            return "absoluteUri"
        case .nfcExternal:
            return "external"
        case .unknown:
            return "unknown"
        case .unchanged:
            return "unchanged"
        @unknown default:
            return "unknown"
        }
    }
}

// MARK: - NFCNDEFReaderSessionDelegate

@available(iOS 13.0, *)
extension NFCModule: NFCNDEFReaderSessionDelegate {
    
    public func readerSession(_ session: NFCNDEFReaderSession, didDetectNDEFs messages: [NFCNDEFMessage]) {
        // 读取模式
        guard !isWriteMode else { return }
        
        var allRecords: [[String: Any]] = []
        for message in messages {
            allRecords.append(contentsOf: parseNDEFMessage(message))
        }
        
        DispatchQueue.main.async {
            // 发送事件
            self.bridge?.sendEvent(BridgeEvent(event: "NFC.TagDetected", data: [
                "records": allRecords
            ]))
            
            // 如果有回调，也调用
            self.pendingCallback?(.success([
                "records": allRecords
            ]))
            self.pendingCallback = nil
        }
    }
    
    public func readerSession(_ session: NFCNDEFReaderSession, didDetect tags: [NFCNDEFTag]) {
        guard let tag = tags.first else { return }
        
        session.connect(to: tag) { error in
            if let error = error {
                session.invalidate(errorMessage: "连接失败: \(error.localizedDescription)")
                return
            }
            
            if self.isWriteMode {
                // 写入模式
                self.writeToTag(tag: tag, session: session)
            } else {
                // 读取模式
                self.readFromTag(tag: tag, session: session)
            }
        }
    }
    
    private func readFromTag(tag: NFCNDEFTag, session: NFCNDEFReaderSession) {
        tag.queryNDEFStatus { status, capacity, error in
            if let error = error {
                session.invalidate(errorMessage: "读取状态失败: \(error.localizedDescription)")
                return
            }
            
            tag.readNDEF { message, error in
                DispatchQueue.main.async {
                    if let error = error {
                        self.pendingCallback?(.failure(BridgeError(
                            code: .internalError,
                            message: "读取失败: \(error.localizedDescription)"
                        )))
                        session.invalidate(errorMessage: "读取失败")
                    } else if let message = message {
                        let records = self.parseNDEFMessage(message)
                        
                        self.bridge?.sendEvent(BridgeEvent(event: "NFC.TagDetected", data: [
                            "records": records,
                            "capacity": capacity
                        ]))
                        
                        self.pendingCallback?(.success([
                            "records": records,
                            "capacity": capacity
                        ]))
                        
                        session.alertMessage = "读取成功"
                        session.invalidate()
                    }
                    
                    self.pendingCallback = nil
                }
            }
        }
    }
    
    private func writeToTag(tag: NFCNDEFTag, session: NFCNDEFReaderSession) {
        guard let message = writeMessage else {
            session.invalidate(errorMessage: "没有要写入的数据")
            return
        }
        
        tag.queryNDEFStatus { status, capacity, error in
            if let error = error {
                session.invalidate(errorMessage: "查询状态失败: \(error.localizedDescription)")
                return
            }
            
            switch status {
            case .notSupported:
                session.invalidate(errorMessage: "标签不支持 NDEF")
                
            case .readOnly:
                session.invalidate(errorMessage: "标签是只读的")
                
            case .readWrite:
                tag.writeNDEF(message) { error in
                    DispatchQueue.main.async {
                        if let error = error {
                            self.pendingCallback?(.failure(BridgeError(
                                code: .internalError,
                                message: "写入失败: \(error.localizedDescription)"
                            )))
                            session.invalidate(errorMessage: "写入失败")
                        } else {
                            self.pendingCallback?(.success([
                                "success": true,
                                "capacity": capacity
                            ]))
                            session.alertMessage = "写入成功"
                            session.invalidate()
                        }
                        
                        self.pendingCallback = nil
                        self.writeMessage = nil
                        self.isWriteMode = false
                    }
                }
                
            @unknown default:
                session.invalidate(errorMessage: "未知状态")
            }
        }
    }
    
    public func readerSession(_ session: NFCNDEFReaderSession, didInvalidateWithError error: Error) {
        DispatchQueue.main.async {
            let nsError = error as NSError
            
            // 用户取消不算错误
            if nsError.code == NFCReaderError.readerSessionInvalidationErrorUserCanceled.rawValue {
                self.pendingCallback?(.success([
                    "cancelled": true
                ]))
            } else if nsError.code != NFCReaderError.readerSessionInvalidationErrorFirstNDEFTagRead.rawValue {
                self.pendingCallback?(.failure(BridgeError(
                    code: .internalError,
                    message: error.localizedDescription
                )))
            }
            
            self.pendingCallback = nil
            self.ndefSession = nil
        }
    }
    
    public func readerSessionDidBecomeActive(_ session: NFCNDEFReaderSession) {
        // 会话激活
        bridge?.sendEvent(BridgeEvent(event: "NFC.SessionActive", data: [:]))
    }
}

File: android/src/main/java/com/aspect/webviewbridge/modules/NFCModule.kt
/**
 * NFC 模块 - 近场通信功能
 *
 * 提供 NFC 读写功能
 */

package com.aspect.webviewbridge.modules

import android.app.Activity
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.nfc.NdefMessage
import android.nfc.NdefRecord
import android.nfc.NfcAdapter
import android.nfc.Tag
import android.nfc.tech.Ndef
import android.nfc.tech.NdefFormatable
import android.os.Build
import android.provider.Settings
import com.aspect.webviewbridge.protocol.BridgeError
import com.aspect.webviewbridge.protocol.BridgeModule
import com.aspect.webviewbridge.protocol.BridgeModuleContext
import com.aspect.webviewbridge.protocol.BridgeRequest
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONArray
import java.nio.charset.Charset
import java.util.Locale

class NFCModule(
    private val context: Context,
    private val bridgeContext: BridgeModuleContext,
    private val activityProvider: () -> Activity?
) : BridgeModule {

    override val moduleName: String = "NFC"

    override val methods: List<String> = listOf(
        "IsAvailable",
        "IsEnabled",
        "StartScan",
        "StopScan",
        "WriteTag",
        "OpenSettings"
    )

    private val nfcAdapter: NfcAdapter? by lazy {
        NfcAdapter.getDefaultAdapter(context)
    }

    private var isScanning = false
    private var isWriteMode = false
    private var pendingWriteMessage: NdefMessage? = null
    private val scope = CoroutineScope(Dispatchers.Main)

    override fun handleRequest(
        method: String,
        request: BridgeRequest,
        callback: (Result<Any?>) -> Unit
    ) {
        when (method) {
            "IsAvailable" -> isAvailable(callback)
            "IsEnabled" -> isEnabled(callback)
            "StartScan" -> startScan(request, callback)
            "StopScan" -> stopScan(callback)
            "WriteTag" -> writeTag(request, callback)
            "OpenSettings" -> openSettings(callback)
            else -> callback(Result.failure(BridgeError.methodNotFound("$moduleName.$method")))
        }
    }

    // MARK: - IsAvailable

    private fun isAvailable(): Map<String, Any> {
        val adapter = nfcAdapter
        return mapOf(
            "isAvailable" to (adapter != null),
            "ndefSupported" to (adapter != null),
            "tagSupported" to (adapter != null)
        )
    }

    // MARK: - IsEnabled

    private fun isEnabled(): Map<String, Boolean> {
        return mapOf(
            "isEnabled" to (nfcAdapter?.isEnabled == true)
        )
    }

    // MARK: - StartScan

    private fun startScan(params: JSONObject): Map<String, Any> {
        val adapter = nfcAdapter
            ?: throw BridgeError(BridgeErrorCode.FeatureDisabled, "NFC 不可用")

        if (!adapter.isEnabled) {
            throw BridgeError(BridgeErrorCode.FeatureDisabled, "NFC 未开启")
        }

        val activity = activityProvider()
            ?: throw BridgeError(BridgeErrorCode.InternalError, "无法获取 Activity")

        isWriteMode = false
        isScanning = true

        enableForegroundDispatch(activity)

        return mapOf(
            "scanning" to true,
            "message" to "NFC 扫描已启动"
        )
    }

    // MARK: - StopScan

    private fun stopScan(): Map<String, Boolean> {
        isScanning = false
        isWriteMode = false
        pendingWriteMessage = null

        activityProvider()?.let { activity ->
            try {
                nfcAdapter?.disableForegroundDispatch(activity)
            } catch (e: Exception) {
                // 忽略
            }
        }

        return mapOf("stopped" to true)
    }

    // MARK: - WriteTag

    private fun writeTag(params: JSONObject): Map<String, Any> {
        val adapter = nfcAdapter
            ?: throw BridgeError(BridgeErrorCode.FeatureDisabled, "NFC 不可用")

        if (!adapter.isEnabled) {
            throw BridgeError(BridgeErrorCode.FeatureDisabled, "NFC 未开启")
        }

        val activity = activityProvider()
            ?: throw BridgeError(BridgeErrorCode.InternalError, "无法获取 Activity")

        // 构建 NDEF 消息
        val records = mutableListOf<NdefRecord>()

        // 写入文本
        params.optString("text", "").takeIf { it.isNotEmpty() }?.let { text ->
            records.add(createTextRecord(text))
        }

        // 写入 URI
        params.optString("uri", "").takeIf { it.isNotEmpty() }?.let { uri ->
            records.add(NdefRecord.createUri(uri))
        }

        // 写入自定义记录
        params.optJSONArray("records")?.let { recordsArray ->
            for (i in 0 until recordsArray.length()) {
                val record = recordsArray.getJSONObject(i)
                createNdefRecord(record)?.let { records.add(it) }
            }
        }

        if (records.isEmpty()) {
            throw BridgeError(BridgeErrorCode.InvalidParams, "text/uri/records")
        }

        pendingWriteMessage = NdefMessage(records.toTypedArray())
        isWriteMode = true
        isScanning = true

        enableForegroundDispatch(activity)

        return mapOf(
            "ready" to true,
            "message" to "请将设备靠近 NFC 标签以写入"
        )
    }

    // MARK: - OpenSettings

    private fun openSettings(): Map<String, Boolean> {
        val intent = Intent(Settings.ACTION_NFC_SETTINGS).apply {
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        context.startActivity(intent)
        return mapOf("opened" to true)
    }

    // MARK: - 前台调度

    private fun enableForegroundDispatch(activity: Activity) {
        val intent = Intent(activity, activity::class.java).apply {
            addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
        }

        val flags = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_MUTABLE
        } else {
            PendingIntent.FLAG_UPDATE_CURRENT
        }

        val pendingIntent = PendingIntent.getActivity(activity, 0, intent, flags)

        val filters = arrayOf(
            IntentFilter(NfcAdapter.ACTION_NDEF_DISCOVERED).apply {
                addCategory(Intent.CATEGORY_DEFAULT)
                try {
                    addDataType("*/*")
                } catch (e: IntentFilter.MalformedMimeTypeException) {
                    throw RuntimeException(e)
                }
            },
            IntentFilter(NfcAdapter.ACTION_TAG_DISCOVERED),
            IntentFilter(NfcAdapter.ACTION_TECH_DISCOVERED)
        )

        val techList = arrayOf(
            arrayOf(Ndef::class.java.name),
            arrayOf(NdefFormatable::class.java.name)
        )

        nfcAdapter?.enableForegroundDispatch(activity, pendingIntent, filters, techList)
    }

    // MARK: - 处理 NFC Intent

    /**
     * 在 Activity 的 onNewIntent 中调用此方法
     */
    fun handleIntent(intent: Intent) {
        if (!isScanning) return

        val action = intent.action ?: return

        when (action) {
            NfcAdapter.ACTION_NDEF_DISCOVERED,
            NfcAdapter.ACTION_TAG_DISCOVERED,
            NfcAdapter.ACTION_TECH_DISCOVERED -> {
                val tag = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    intent.getParcelableExtra(NfcAdapter.EXTRA_TAG, Tag::class.java)
                } else {
                    @Suppress("DEPRECATION")
                    intent.getParcelableExtra(NfcAdapter.EXTRA_TAG)
                }

                tag?.let {
                    if (isWriteMode) {
                        handleWrite(it)
                    } else {
                        handleRead(it, intent)
                    }
                }
            }
        }
    }

    // MARK: - 读取

    private fun handleRead(tag: Tag, intent: Intent) {
        val ndef = Ndef.get(tag)

        if (ndef != null) {
            try {
                ndef.connect()
                val message = ndef.cachedNdefMessage ?: ndef.ndefMessage

                if (message != null) {
                    val records = parseNdefMessage(message)

                    scope.launch {
                        bridgeContext.sendEvent("NFC.TagDetected", mapOf(
                            "records" to records,
                            "capacity" to ndef.maxSize,
                            "isWritable" to ndef.isWritable
                        )))
                    }
                }
            } catch (e: Exception) {
                scope.launch {
                    bridgeContext.sendEvent("NFC.Error", mapOf(
                        "error" to (e.message ?: "读取失败")
                    )))
                }
            } finally {
                try {
                    ndef.close()
                } catch (e: Exception) {
                    // 忽略
                }
            }
        } else {
            // 尝试从 Intent 读取
            val messages = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                intent.getParcelableArrayExtra(NfcAdapter.EXTRA_NDEF_MESSAGES, NdefMessage::class.java)
            } else {
                @Suppress("DEPRECATION")
                intent.getParcelableArrayExtra(NfcAdapter.EXTRA_NDEF_MESSAGES)
            }

            messages?.let { msgs ->
                val allRecords = mutableListOf<Map<String, Any?>>()
                msgs.filterIsInstance<NdefMessage>().forEach { message ->
                    allRecords.addAll(parseNdefMessage(message))
                }

                scope.launch {
                    bridgeContext.sendEvent("NFC.TagDetected", mapOf(
                        "records" to allRecords
                    ))
                }
            }
        }
    }

    // MARK: - 写入

    private fun handleWrite(tag: Tag) {
        val message = pendingWriteMessage ?: return

        val ndef = Ndef.get(tag)
        if (ndef != null) {
            try {
                ndef.connect()

                if (!ndef.isWritable) {
                    sendWriteError("标签是只读的")
                    return
                }

                if (ndef.maxSize < message.byteArrayLength) {
                    sendWriteError("数据太大，超出标签容量")
                    return
                }

                ndef.writeNdefMessage(message)

                scope.launch {
                    bridgeContext.sendEvent("NFC.WriteSuccess", mapOf(
                        "success" to true,
                        "capacity" to ndef.maxSize
                    ))
                }

                // 写入成功后重置状态
                isWriteMode = false
                pendingWriteMessage = null

            } catch (e: Exception) {
                sendWriteError(e.message ?: "写入失败")
            } finally {
                try {
                    ndef.close()
                } catch (e: Exception) {
                    // 忽略
                }
            }
        } else {
            // 尝试格式化
            val formatable = NdefFormatable.get(tag)
            if (formatable != null) {
                try {
                    formatable.connect()
                    formatable.format(message)

                    scope.launch {
                        bridgeContext.sendEvent("NFC.WriteSuccess", mapOf(
                            "success" to true,
                            "formatted" to true
                        )))
                    }

                    isWriteMode = false
                    pendingWriteMessage = null

                } catch (e: Exception) {
                    sendWriteError(e.message ?: "格式化失败")
                } finally {
                    try {
                        formatable.close()
                    } catch (e: Exception) {
                        // 忽略
                    }
                }
            } else {
                sendWriteError("标签不支持 NDEF")
            }
        }
    }

    private fun sendWriteError(message: String) {
        scope.launch {
            bridgeContext.sendEvent("NFC.WriteError", mapOf(
                "success" to false,
                "error" to message
            )))
        }
    }

    // MARK: - 辅助方法

    private fun createTextRecord(text: String): NdefRecord {
        val lang = Locale.getDefault().language
        val langBytes = lang.toByteArray(Charset.forName("US-ASCII"))
        val textBytes = text.toByteArray(Charset.forName("UTF-8"))
        val payload = ByteArray(1 + langBytes.size + textBytes.size)

        payload[0] = langBytes.size.toByte()
        System.arraycopy(langBytes, 0, payload, 1, langBytes.size)
        System.arraycopy(textBytes, 0, payload, 1 + langBytes.size, textBytes.size)

        return NdefRecord(NdefRecord.TNF_WELL_KNOWN, NdefRecord.RTD_TEXT, ByteArray(0), payload)
    }

    private fun createNdefRecord(record: JSONObject): NdefRecord? {
        val tnf = when (record.optString("tnf", "wellKnown")) {
            "empty" -> NdefRecord.TNF_EMPTY
            "wellKnown" -> NdefRecord.TNF_WELL_KNOWN
            "media" -> NdefRecord.TNF_MIME_MEDIA
            "absoluteUri" -> NdefRecord.TNF_ABSOLUTE_URI
            "external" -> NdefRecord.TNF_EXTERNAL_TYPE
            else -> NdefRecord.TNF_UNKNOWN
        }

        val type = record.optString("type", "").toByteArray()
        val id = record.optString("id", "").toByteArray()
        val payload = record.optString("payload", "").toByteArray()

        return NdefRecord(tnf.toShort(), type, id, payload)
    }

    private fun parseNdefMessage(message: NdefMessage): List<Map<String, Any?>> {
        return message.records.map { record ->
            val info = mutableMapOf<String, Any?>(
                "tnf" to tnfString(record.tnf),
                "type" to String(record.type, Charset.forName("UTF-8")),
                "id" to String(record.id, Charset.forName("UTF-8")),
                "payload" to android.util.Base64.encodeToString(record.payload, android.util.Base64.NO_WRAP)
            )

            // 解析文本
            if (record.tnf == NdefRecord.TNF_WELL_KNOWN &&
                record.type.contentEquals(NdefRecord.RTD_TEXT)) {
                parseTextRecord(record)?.let { text ->
                    info["text"] = text
                }
            }

            // 解析 URI
            if (record.tnf == NdefRecord.TNF_WELL_KNOWN &&
                record.type.contentEquals(NdefRecord.RTD_URI)) {
                val uri = record.toUri()
                info["uri"] = uri?.toString()
            }

            info
        }
    }

    private fun parseTextRecord(record: NdefRecord): String? {
        val payload = record.payload
        if (payload.isEmpty()) return null

        val languageCodeLength = payload[0].toInt() and 0x3F
        if (payload.size <= languageCodeLength + 1) return null

        return String(
            payload,
            languageCodeLength + 1,
            payload.size - languageCodeLength - 1,
            Charset.forName("UTF-8")
        )
    }

    private fun tnfString(tnf: Short): String {
        return when (tnf) {
            NdefRecord.TNF_EMPTY -> "empty"
            NdefRecord.TNF_WELL_KNOWN -> "wellKnown"
            NdefRecord.TNF_MIME_MEDIA -> "media"
            NdefRecord.TNF_ABSOLUTE_URI -> "absoluteUri"
            NdefRecord.TNF_EXTERNAL_TYPE -> "external"
            NdefRecord.TNF_UNKNOWN -> "unknown"
            NdefRecord.TNF_UNCHANGED -> "unchanged"
            else -> "unknown"
        }
    }

    /**
     * 清理资源
     */
    fun cleanup() {
        stopScan()
    }
}
